Adding gemqrt function.

TeachRaccooon opened the pull request:
Adding a gemqrt() routine, needed in RandLAPACK.
Code is initially generated via `lapackpp/tools/wrapper_gen.py`.

mgates3 commented on test/test_gemqrt.cc:
```diff
@@ -0,0 +1,189 @@
+
+//------------------------------------------------------------------------------
+// Simple overloaded wrappers around LAPACKE (assuming routines in LAPACKE).
+// These should go in test/lapacke_wrappers.hh.
```
Can you move these 4 wrappers to lapacke_wrappers.hh?

TeachRaccooon replied:
Done.

mgates3 commented on test/test.cc:
```diff
@@ -251,6 +251,7 @@ std::vector< testsweeper::routines_t > routines = {
     { "gelqf",              test_gelqf,     Section::qr }, // tested numerically
     { "geqlf",              test_geqlf,     Section::qr }, // tested numerically
     { "gerqf",              test_gerqf,     Section::qr }, // tested numerically; R, Q are full sizeof(A), could be smaller
+    { "gemqrt",             test_gemqrt,    Section::qr }, // tested numerically
```
Change to "tested via LAPACKE".
Sorry the comments here are unclear. "tested numerically" means it has a backwards error check, while "tested via LAPACKE" means the result is compared to LAPACKE (a forward error check).

TeachRaccooon replied:
Fixed.

mgates3 commented on test/test_gemqrt.cc:
```diff
@@ -0,0 +1,189 @@
+    lapack::larnv( idist, iseed, T.size(), &T[0] );
+    lapack::larnv( idist, iseed, C_tst.size(), &C_tst[0] );
+    C_ref = C_tst;
+
```
You will need to add a call here to `lapack::geqrt` before calling `gemqrt` in order to setup the matrices, before running the test.

TeachRaccooon replied:
Got it, fixed. 
I still have the actual test call commented out; it is likely failing because i am setting up the line in run_test.py wrong.

TeachRaccooon replied:
Fixed.

mgates3 reviewed (changes requested):
Not sure what's causing it to segfault in CI. Will need `geqrt` in tester, but omitting that shouldn't cause segfaults, it seems.

TeachRaccooon commented:
> Not sure what's causing it to segfault in CI. Will need `geqrt` in tester, but omitting that shouldn't cause segfaults, it seems.

I was thinking that that is due to the wrong setup of the line in run_tests.py, but seems like something else is the issue.
Also, I noticed that there are no tests for gemqr.

TeachRaccooon commented:
Just tested `gemqrt` in RandLAPACK and the routine itself works fine.
Looks like something is off with the way I've set up tests.

TeachRaccooon commented:
Hi @mgates3 
I see that there is apending change request, but I think I addressed all the issues that were there (unless github is not showing me something for whatever reason).
So just wanted to check if we can merge this PR.

mgates3 commented:
@TeachRaccooon That just means I haven't approved the changes. I'll look at it probably today or tomorrow.

mgates3 commented on test/test_gemqrt.cc:
```diff
@@ -0,0 +1,139 @@
+
+    // Calling this to set up the matrices
+    lapack::geqrt( m, n, nb, &V[0], ldv, &T[0], ldt );
+    lapack::gemqrt( side, trans, m, n, k, nb, &V[0], ldv, &T[0], ldt, &C_tst[0], ldc );
```
Testing the code is required, i.e., remove this `gemqrt` above and enable the code below. I'm looking into what the issues are — it seems some issue with LAPACKE returning errors.

mgates3 commented:
I noticed your fork is specific to this feature (`TeachRaccooon/lapackpp-add_gemqrt`) and uses the `master` branch in that fork. That works, but it's usually cleaner to create one fork (say, `TeachRaccooon/lapackpp`) and create a branch for each specific feature (say, branch `gemqrt`).

mgates3 commented:
Can you add an empty commit to trigger the CI actions again? The "re-run jobs" button will re-run with the old master, not with the latest master, which seems odd to me.
```
git commit --allow-empty -m "Empty-Commit"
```
I'm going to squash all this into one commit when merging, so the empty commit will disappear.
